src/main/java/org/jetbrains/bsp/bazel/server/DiagnosticsDispatcher.java

```diff
@@ -0,0 +1,139 @@
+        Diagnostics.TargetDiagnostics.parseFrom(Files.readAllBytes(Paths.get(diagnosticsLocation)));
+
+    for (Diagnostics.FileDiagnostics fileDiagnostics : targetDiagnostics.getDiagnosticsList()) {
+      LOGGER.info("Inserting diagnostics for path: " + fileDiagnostics.getPath());
+
+      filesToDiagnostics.put(
+          getUriForPath(fileDiagnostics.getPath()), convertDiagnostics(target, fileDiagnostics));
+    }
+  }
+
+  public void emitDiagnostics(
+      Map<Uri, List<PublishDiagnosticsParams>> filesToDiagnostics, BuildTargetIdentifier target) {
+    for (SourceItem source : bspServer.getCachedBuildTargetSources(target)) {
+      Uri sourceUri = Uri.fromFileUri(source.getUri());
+      if (!filesToDiagnostics.containsKey(sourceUri)) {
+        addSource(filesToDiagnostics, sourceUri, target);
+      }
+
+      if (bspClient != null) {
+        for (List<PublishDiagnosticsParams> values : filesToDiagnostics.values()) {
+          for (PublishDiagnosticsParams param : values) {
+            bspClient.onBuildPublishDiagnostics(param);
+          }
+        }
+      }
+    }
+  }
+
+  private void addSource(
+      Map<Uri, List<PublishDiagnosticsParams>> filesToDiagnostics,
+      Uri sourceUri,
+      BuildTargetIdentifier target) {
+    PublishDiagnosticsParams publishDiagnosticsParams =
+        new PublishDiagnosticsParams(
+            new TextDocumentIdentifier(sourceUri.toString()), target, new ArrayList<>(), true);
+
+    filesToDiagnostics.put(sourceUri, Lists.newArrayList(publishDiagnosticsParams));
+  }
+
+  private List<PublishDiagnosticsParams> convertDiagnostics(
+      BuildTargetIdentifier target, Diagnostics.FileDiagnostics request) {
+    List<Diagnostic> diagnostics =
+        request.getDiagnosticsList().stream()
+            .map(this::convertDiagnostic)
+            .collect(Collectors.toList());
+
+    PublishDiagnosticsParams publishDiagnosticsParams =
+        new PublishDiagnosticsParams(
+            new TextDocumentIdentifier(getUriForPath(request.getPath()).toString()),
+            target,
+            diagnostics,
+            true);
+
+    return Lists.newArrayList(publishDiagnosticsParams);
+  }
+
+  private Diagnostic convertDiagnostic(Diagnostics.Diagnostic diagProto) {
+    DiagnosticSeverity severity = convertSeverity(diagProto.getSeverity());
+
+    Position startPosition =
+        new Position(
+            diagProto.getRange().getStart().getLine(),
+            diagProto.getRange().getStart().getCharacter());
+    Position endPosition =
+        new Position(
+            diagProto.getRange().getEnd().getLine(), diagProto.getRange().getEnd().getCharacter());
+    Range range = new Range(startPosition, endPosition);
+
+    Diagnostic diagnostic = new Diagnostic(range, diagProto.getMessage());
+    if (severity != null) {
+      diagnostic.setSeverity(severity);
+    }
+
+    return diagnostic;
+  }
+
+  private DiagnosticSeverity convertSeverity(Diagnostics.Severity protoSeverity) {
```

Comment:
Accidentally  deleted previous comment, but there is my answer:
it is a simple mapping `Diagnostics.Severity` -> `DiagnosticSeverity` what means that it can be represented by `Map<Diagnostics.Severity, DiagnosticSeverity>`

Reply:
Right, good idea